### DISCUSSION TITLE
How to add background processes to server (e.g. Discord Bot)?

### DISCUSSION THREAD

**Discussion by masterjanic** [nuxt/nuxt#17682] [Questions]
Hey, I have a more like general question regarding the project structure. 

I want to use the api routes system of Nuxt to create my api and retrieve data from my database, but I need to implement an interval task that loops through the database entries to perform checks.

I also want to integrate a running Discord bot into my project. In older version of my projects I had a separate repository for my server side (express server) where such functions were running. 

**Is there any possibility to implement these kind of background tasks / processes directly into a single Nuxt repository? Or do I need a custom server or even a separated project?**

Thank you in advance!

**Comment by benoitdemaegdt** [comment]
Hey @masterjanic  👋 

As far as I know, there is no such thing in Nuxt.

However, if deploying your application on Netlify is an option, you can have a look at [backgroud functions](https://docs.netlify.com/functions/background-functions/).

Hope it will help 😉

**Comment by TheAlexLichter** [comment]
I suggest using a standalone application for e.g. a Discord bot to:

* Reduce the single point of failure
* Scale independently (the webapp and your Discord bot)

**Comment by masterjanic** [comment]
@benoitdemaegdt Thanks for you answer, but I don't want to use Netlify, since I want to host my application on a private server.

@manniL I unterstand your point of view, but it is a rather small project and the Discord bot integrates with the functionality of the database and web app. Therefore, I am looking for a solution to bundle both in one project without the need of two repositories. Is there any other solution to this problem?

**Comment by pi0** [comment]
You can use a [Nitro Plugin](https://nitro.unjs.io/guide/plugins.html) to initiate background process by putting it in `server/plugins`.

**Reply by JavascriptMick** [reply]
like this?
```
//plugins/1.background_processing.server.ts
export default defineNuxtPlugin(nuxtApp => {
  console.log('initialising worker');
  startWork();
});

const startWork = () => setTimeout(work, 5000);

const work = async () => {
  try {
    //Your work goes here but for demonstration purposes, do fake work
    console.log('Doing Work');
    await fakeWorkDelayDontUseThis(2000);
  } catch (e) {
    console.log(`Error occured while doing work ${e}`);
  } finally {
    console.log('work done, re-initialising worker');
    startWork();
  }
};

const fakeWorkDelayDontUseThis = (ms: number): Promise<void> => {
  return new Promise(resolve => setTimeout(resolve, ms));
};
```

**Reply by JavascriptMick** [reply]
After some time I realise that this method really sucks bad.  e.g. I need to hit api's on the Nuxt site in order to 'start' the background process and it is pretty flakey.  Not Recommending this method.

**Reply by TheAlexLichter** [reply]
@JavascriptMick That is only the case if you deploy on edge/serverless where functions are started on demand I think

**Reply by JavascriptMick** [reply]
@manniL in practice the 'hit the api' problem even occurs when running locally. Also, pretty sure hot reload spins up a second instance with the old instance (with old code) still running which is super annoying.

**Comment by JavascriptMick** [comment]
I had another crack at this by using Vite to compile a standalone node server (https://stackblitz.com/edit/github-jpamey?file=background%2Fsome.server.ts) it works for simple cases but gets shitty when you use a lot of libraries in particular Prisma.

**Reply by JavascriptMick** [reply]
I have tweaked the above example to use vite-node in preference to building the module and then running with node.  It's a lot more stable when using a lot of libraries.

**Comment by ongnxco** [comment]
> I have tweaked the above example to use vite-node in preference to building the module and then running with node. It's a lot more stable when using a lot of libraries.
So that means I can't put the discord bot code in the server folder?
